runtime/src/accounts_db.rs

```diff
@@ -151,6 +151,87 @@ pub enum IncludeSlotInHash {
     IrrelevantAssertOnUse,
 }
 
+#[derive(Default)]
+/// hold alive accounts and bytes used by those accounts
+/// alive means in the accounts index
+struct AliveAccounts<'a> {
+    accounts: Vec<&'a StoredAccountMeta<'a>>,
+    bytes: usize,
+}
+
+/// separate pubkeys into those with a single refcount and those with > 1 refcount
+struct ShrinkCollectAliveSeparatedByRefs<'a> {
+    one_ref: AliveAccounts<'a>,
+    many_refs: AliveAccounts<'a>,
+}
+
+trait ShrinkCollectRefs<'a>: Sync + Send {
```

Comment:
If users of a trait require sync/send, then I have found it convenient to just put the required traits on the trait instead of putting it on multiple functions in a call chain. It is true that some users of this trait at some point in the future may not require sync/send and that could preclude an impl of some type, but I have no expectation of that occurring. And if it does, we could move sync/send. wdyt?